design/20210324.certificate-request-approve-deny.md

```diff
@@ -0,0 +1,191 @@
+sufficient RBAC permissions. These permissions are based upon the request
+itself- specifically the request's IssuerRef:
+
+```yaml
+apiGroups: ["cert-manager.io"]
+resources: ["signers"]
+verbs: ["approve"]
+resourceNames: ["<issuer-kind>.<issuer-group>/<issuer-name or * for all issuer
+names>"]
+```
+
+An example ClusterRole that would grant the permissions to set the Approve and
+Denied conditions of CertificateRequests that reference the `MyIssuer` external
+issuer, in the group `my-example.io`, with the name `MyApp`:
+
+```
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: my-example-io-my-issuer-myapp-approver
+rules:
+  - apiGroups: ["cert-manager.io"]
+    resources: ["signers"]
+    verbs: ["approve"]
+    resourceNames: ["myissuer.my-example.io/MyApp"]
+```
+
+These permissions are enforced using a
+[SubjectAccessReview](https://github.com/kubernetes/kubernetes/blob/b11d0fbdd58394a62622787b38e98a620df82750/pkg/apis/authorization/types.go#L27)
+that is performed in the cert-manager Webhook component when a user attempts to
+Approve or Deny a request. The user of the approver is tested along with the
+IssuerRef of the CertificateRequest.
+
+### Scope
+The RBAC permissions *must* be granted at the cluster scope (i.e.
+ClusterRoleBinding). This matches the same scoping behaviour as
+[certificates.k8s.io
+CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).
+
+### API Group
+The apiGroup must *always* be `cert-manage.io`, as all CertificateRequests are
+in the `cert-manager.io` group.
+
+### Resource
+The resource must *always* be `signers`. This matches the same resource as
+[required by certificates.k8s.io
+CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).
+
+### Verbs
+The verb must *always* be `approve`. This verb grants the approver the
+permissions to set *both* Approved and Denied conditions.
+
+### Resource Names
+The resource name is made up of 3 parts: the issuer kind, issuer group, and the
+issuer name.
+
+The issuer kind must be lower case, i.e. "ClusterIssuer" would become
+"clusterissuer".
+
+The issuer group should be the same string of the group of issuer.
+
+The name may be either a wild card '*' denoting all issuer names, or instead a
+string which must match the issuer name as it appears in the IssuerRef.
+
+An example of signing all `MyIssuer` issuers, and `ClusterMyIssuers` with the name
+`MyApp`, in the `my-example.io` group:
+
+```yaml
+    resourceNames: ["myissuer.my-example.io/*", "clustermyissuer.my-example.io/MyApp"]
+```
+
+## cert-manager Approver
+
+The cert-manager controller is deployed with an internal approver controller.
+This controller will attempt to Approve *all* CertificateRequests that are
+created, regardless of its contents or IssuerRef.
```

Comment:
Do we want the internal controller to approve all *external* ones too? How will authZ errors be handled?

Reply:
Currently it will just error out against the rate limiting queue from the validation failure. How could we improve this?

I think we do want the approver to retry, since an administrator may genuinely have not set the right RBAC up yet.

I don't think we want the approver to have RBAC for all issuers there could be.. This makes the story of introducing other approvers more tricky, and it not really possible to represent anyway.